File: 2019/day2/problem.ts
import { solve, TestCase, ProblemInput } from '../../lib/typescript'

const part1TestCases: TestCase[] = [
  {
    input: '1,9,10,3,2,3,11,0,99,30,40,50',
    answer: 3500,
  },
]
function part1Solution({ input, testCase }: ProblemInput): number {
  const intcode = input.split(',').map(Number)
  if (!testCase) {
    intcode[1] = 12
    intcode[2] = 2
  }

  for (let x = 0; x < intcode.length; x += 4) {
    const opcode = intcode[x]
    if (opcode === 1) {
      intcode[intcode[x + 3]] = intcode[intcode[x + 1]] + intcode[intcode[x + 2]]
    } else if (opcode === 2) {
      intcode[intcode[x + 3]] = intcode[intcode[x + 1]] * intcode[intcode[x + 2]]
    } else if (opcode === 99) {
      break
    }
  }
  return intcode[0]
}

const part2TestCases: TestCase[] = [
  {
    input: '1,0,0,0,99',
    answer: 2,
  },
]
function part2Solution({ input }: ProblemInput): number {
  return 0
}

solve({ input: `${__dirname}/input.txt`, part1TestCases, part1Solution, part2TestCases, part2Solution })

File: 2023/day8/problem.rb
require 'byebug'
require_relative '../../lib/graph'
require_relative '../../lib/array'

class Solution < AbstractSolution
  def initialize
    super
    @part1_test_cases = [
      {
        answer: 2,
        input: <<~TXT
          RL

          AAA = (BBB, CCC)
          BBB = (DDD, EEE)
          CCC = (ZZZ, GGG)
          DDD = (DDD, DDD)
          EEE = (EEE, EEE)
          GGG = (GGG, GGG)
          ZZZ = (ZZZ, ZZZ)
        TXT
      }
    ]

    @part2_test_cases = [
      {
        answer: 6,
        input: <<~TXT
          LR

          11A = (11B, XXX)
          11B = (XXX, 11Z)
          11Z = (11B, XXX)
          22A = (22B, XXX)
          22B = (22C, 22C)
          22C = (22Z, 22Z)
          22Z = (22B, 22B)
          XXX = (XXX, XXX)
        TXT
      }
    ]
  end

  def parse_graph(input)
    pattern = input.split("\n").first
    graph = Graph.new
    input.split("\n")[2..].each do |node|
      name, left, right = node.scan(/(\w+) = \((\w+), (\w+)\)/).first
      graph.add_edge(from: name, to: left, weight: 1) if name != left
      graph.add_edge(from: name, to: right, weight: 1) if name != right
    end
    return pattern, graph
  end

  def part1(input)
    pattern, graph = parse_graph(input)
    return graph.shortest_path(from: 'AAA', to: 'ZZZ')[1].lcm(pattern.length)
  end

  def part2(input)
    pattern, graph = parse_graph(input)
    start_nodes = graph.nodes.select { _1.end_with?('A') }
    end_nodes = graph.nodes.select { _1.end_with?('Z') }
    start_nodes.reduce(pattern.length) do |answer, start_node|
      end_nodes.each do |end_node|
        shortest_path = graph.shortest_path(from: start_node, to: end_node)
        answer = answer.lcm(shortest_path[1]) if shortest_path
      end
      answer
    end
  end
end
